sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java

```diff
@@ -0,0 +1,214 @@
+import org.apache.beam.sdk.util.TimerSpec;
+import org.apache.beam.sdk.util.TimerSpecs;
+import org.apache.beam.sdk.util.state.BagState;
+import org.apache.beam.sdk.util.state.StateSpec;
+import org.apache.beam.sdk.util.state.StateSpecs;
+import org.apache.beam.sdk.util.state.ValueState;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link PTransform} that allows to compute elements in batch of desired size. The input {@link
+ * PCollection} needs to be a {@code PCollection<KV>}. Elements that must belong to the same batch
+ * need to have the same key. Elements are added to a buffer. When the buffer reaches {@code
+ * batchSize}, it is then processed through a user {@link SimpleFunction perBatchFn} function. The
+ * output elements then are added to the output {@link PCollection}. Windows are preserved (batches
+ * contain elements from the same window). Batching is done trans-bundles (batches may contain
+ * elements from more than one bundle)
+ *
+ * <p>Example (batch call a webservice and get return codes)
+ *
+ * <pre>{@code
+ * SimpleFunction<Iterable<String>, Iterable<String>> perBatchFn =
+ * new SimpleFunction<Iterable<String>, Iterable<String>>() {
+ *   {@literal @}Override
+ *   public Iterable<String> apply(Iterable<String> input) {
+ *     ArrayList<String> results = callWebService(input);
+ *     return results;
+ *     }
+ *  };
+ *  ...
+ *  Pipeline pipeline = Pipeline.create(...);
+ *  ...
+ *  long batchSize = 100L;
+ *  pipeline.apply(BatchingParDo.via(batchSize, perBatchFn))
+ *          .setCoder(StringUtf8Coder.of());
+ *  pipeline.run();
+ * }</pre>
+ * *
+ */
+public class BatchingParDo<K, InputT, OutputT>
+    extends PTransform<PCollection<KV<K, InputT>>, PCollection<KV<K, OutputT>>> {
+
+  private final long batchSize;
+  private final SimpleFunction<? super Iterable<InputT>, ? extends Iterable<OutputT>> perBatchFn;
+
+  private BatchingParDo(
+      long batchSize,
+      SimpleFunction<? super Iterable<InputT>, ? extends Iterable<OutputT>> perBatchFn) {
+    this.batchSize = batchSize;
+    this.perBatchFn = perBatchFn;
+  }
+
+  public static <K, InputT, OutputT> BatchingParDo<K, InputT, OutputT> via(
+      long batchSize,
+      SimpleFunction<? super Iterable<InputT>, ? extends Iterable<OutputT>> perBatchFn) {
+    return new BatchingParDo<>(batchSize, perBatchFn);
+  }
+
+  @Override
+  public PCollection<KV<K, OutputT>> expand(PCollection<KV<K, InputT>> input) {
+    Duration allowedLateness = input.getWindowingStrategy().getAllowedLateness();
+
+    PCollection<KV<K, OutputT>> output =
+        input.apply(
+            ParDo.of(
+                new BatchingDoFn<>(
+                    batchSize,
+                    perBatchFn,
+                    allowedLateness,
+                    (Coder<InputT>) input.getCoder().getCoderArguments().get(1),
```

Comment:
I would prefer to `checkState` that the input coder is a `KvCoder`, cast to it, and explicitly extract the `keyCoder` and `valueCoder` from it. Any failure would be much simpler to debug.